### PR TITLE
Portfolio gallery: extract composable and test EXIF formatting

### LINKED ISSUE #219: Portfolio gallery: extract composable and test EXIF formatting
## Parent

#211

## Source review

- Codebase architecture review (June 2026)
- Tag: `[architecture]`

---

## Problem Statement

The home route **photo gallery** orchestrates loading, masonry layout prep, hover **photo details**, **photo preview** (zoom, pan, back navigation), and EXIF parsing inside a single page component (~650 lines). The **portfolio shell** should stay thin; this is the main static-page orchestration hotspot outside **projects**.

Separately, the pure formatting layer that turns parsed metadata into hover summaries and **photo preview** panel rows has no unit tests, so regressions in **photo details** formatting would only show up through manual browsing.

## Solution

Extract gallery orchestration into a dedicated composable so the home page SFC owns layout and wiring only. Add unit tests at the existing pure-formatting module boundary. All visitor-facing behavior stays unchanged—this is housekeeping for maintainability, not a product change.

## User Stories

1. As a site maintainer, I want **photo gallery** loading, hover **photo details**, and **photo preview** logic outside the home page SFC, so that the **portfolio shell** page stays readable and changes are localized.
2. As a site maintainer, I want **gallery loading** progress and masonry ratio measurement encapsulated in one composable, so that the blocking gate behavior (full grid after all ratios are ready) is easy to find and modify later.
3. As a site maintainer, I want hover **photo details** fetch and cache behavior colocated with **photo preview** EXIF loading, so that metadata parsing is not duplicated across handlers.
4. As a visitor browsing the **photo gallery**, I want the masonry grid to appear exactly as today after **gallery loading** completes, so that the refactor does not change my browsing experience.
5. As a visitor, I want tile hover to show the same compact **photo details** summary as today, so that camera and exposure info remains discoverable on desktop.
6. As a visitor, I want clicking a tile to open **photo preview** with the same image, toolbar, and panel layout as today, so that inspecting photos feels unchanged.
7. As a visitor, I want **photo preview zoom** to work as today (fit by default, click-to-zoom-at-point, pan when zoomed, toolbar toggle), so that inspection gestures are preserved.
8. As a visitor, I want **photo preview back navigation** to dismiss the overlay and restore **photo gallery** scroll position, so that browser Back behaves like closing a stack layer—not like **project shell** back trapping.
9. As a visitor with a photo that has **missing photo details**, I want the image to display without error messaging, so that stripped or unreadable metadata never blocks the gallery.
10. As a visitor, I want GPS/location metadata never surfaced in **photo details**, so that location privacy expectations are unchanged.
11. As a developer, I want unit tests on the pure formatting functions that produce hover lines and preview panel rows, so that representative metadata shapes and edge cases are guarded without brittle UI tests.
12. As a developer, I want tests to assert structured outputs (line arrays, label/value rows), not rendered copy strings, so that formatting tweaks do not require snapshot churn.
13. As a developer working on the thermonuclear-clean pass (#211), I want the home route to follow the same composable-extraction pattern used elsewhere (orchestration out of SFCs), so that static portfolio pages match project code-quality expectations.
14. As a developer, I want ESLint clean on all touched files, so that CI stays green.
15. As a maintainer cloning a repo with no checked-in photos, I want the **empty gallery fallback** (setup instructions and Google Photos embed) to behave as today, so that local dev ergonomics are unchanged.

## Implementation Decisions

- **Behavior freeze:** No visitor-facing UI/UX changes. Every resolved glossary behavior (**gallery loading**, **photo preview**, **photo preview zoom**, **photo preview back navigation**, **photo details**, **missing photo details**, **empty gallery fallback**) must match production today.
- **Composable boundary:** Introduce `usePhotoGallery()` (or equivalent name) owning:
  - Photo catalog preparation (checked-in assets → sorted tile descriptors with thumb URLs)
  - **Gallery loading** state (progress, loaded count, ready flag, aspect-ratio measurement)
  - EXIF load + in-memory cache (still via `exifr`; GPS may be parsed but must not reach formatting output shown to visitors)
  - Hover **photo details** state (active tile, summary lines)
  - **Photo preview** state (open/close, src/label, **photo details** panel rows and toggle, zoom/pan coordination with existing pointer-drag helper)
  - **Photo preview back navigation** (history push/pop, scroll restore, close-via-X sync with `history.back()`)
- **Page SFC role:** Home page retains template markup, scoped styles (masonry columns, preview dialog layout, transitions), and binds to composable refs/handlers. Asset glob wiring may live in the composable or a small adjacent module—whichever keeps the SFC thinnest without changing behavior.
- **Formatting module:** Keep the existing pure formatting module responsible for turning normalized EXIF objects into hover summary lines and preview panel rows. Do not merge `exifr` parsing into this module—it stays presentation-only.
- **Reuse existing helpers:** Continue using the shared pointer-drag-scroll composable for **photo preview zoom** panning; do not reimplement drag/pan logic.
- **No new product surfaces:** Do not introduce filtering, albums, reordering UI, or changes to masonry breakpoints.

## Testing Decisions

**What makes a good test here:** Assert pure function inputs/outputs and stable data shapes—never DOM text, aria labels, or CSS classes. Tests should fail only when formatting logic or contracts change, not when copy or styling shifts.

**Primary test seam (required):** The pure formatting module exports (`pickExifCore`, `formatExifTimestamp`, `exifSummaryLines`, `exifToRows`). Feed representative fixture objects mimicking normalized `exifr` output—full metadata, partial fields, empty object, null/non-object input, alternate key casings, fractional exposure arrays, malformed dates.

**Prior art:** Node `node:test` + `node:assert/strict` colocated `*.test.js` files (e.g. composable controller tests under `src/composables/` and `src/layouts/projects/composables/`). Follow the same style—no Vue Test Utils for this slice.

**Out of test scope for this slice:**
- Composable/Vue integration tests (no new seam required; behavior verified by visual QA on the home route)
- `exifr` network/binary parsing (integration concern; composable keeps current parse options)
- End-to-end or snapshot tests of the masonry grid

**Visual QA checklist (manual):**
- **Gallery loading** progress then full grid
- Hover **photo details** on a tile with metadata
- Open **photo preview**, toggle **photo details** panel, exercise **photo preview zoom**
- Browser Back closes **photo preview** and restores scroll
- Tile with **missing photo details** still opens preview without error UI

## Out of Scope

- New **photo gallery** features (filtering, albums, sort controls)
- Changes to checked-in photo assets or thumbnail generation scripts
- **Share metadata catalog** / paste-unfurl work
- Reordering photos, masonry breakpoint changes, or Google Photos fallback copy changes
- GPS surfacing in **photo details**
- Unit tests for the composable itself (unless a natural pure controller seam emerges during extraction— not a goal of this slice)

## Further Notes

- Domain glossary terms for this surface live in root `CONTEXT.md` (**photo gallery**, **gallery loading**, **photo preview**, **photo details**, **missing photo details**, **photo preview back navigation**, **photo preview zoom**, **empty gallery fallback**).
- Parent epic #211 tracks the broader thermonuclear-clean architecture pass; this slice is the portfolio static-page hotspot called out in the June 2026 review.
- Acceptance criteria:
  - [ ] Gallery loading / hover / **photo preview** logic lives outside the home page SFC in a composable or feature module
  - [ ] Pure formatting module has unit tests for representative metadata shapes and edge cases
  - [ ] Home route **photo gallery** behavior unchanged (visual QA)
  - [ ] ESLint clean on touched files

### PR DESCRIPTION
**Pull request by enmaku**
## Summary

- Extract photo gallery orchestration from `IndexPage.vue` into `usePhotoGallery()` (loading, hover photo details, photo preview, back navigation, zoom/pan).
- Add unit tests for the pure EXIF formatting module (`pickExifCore`, `formatExifTimestamp`, `exifSummaryLines`, `exifToRows`).
- Wire `src/utils/**/*.test.js` into the npm test script.

Closes #219.

## Test plan

- [x] `npm test` — new EXIF formatting tests pass
- [x] ESLint clean on touched files
- [x] Manual visual QA on home route:
  - [x] Gallery loading progress → full masonry grid
  - [x] Hover photo details on a metadata-rich tile
  - [x] Photo preview: details panel toggle, zoom/pan, toolbar
  - [x] Browser Back closes preview and restores scroll
  - [x] Tile with missing photo details — no error UI
  - [x] Empty gallery fallback unchanged (no local photos)

### CODE REVIEW

File: src/utils/exifFormat.test.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    ...fullExifFixture,
+    Make: 'FUJIFILM',
+    DateTimeOriginal: '2024:06:10 14:30:00',
+    ExifImageWidth: 6240,
+    ExifImageHeight: 4160,
+  })
+
+  const labels = rows.map((row) => row.label)
+
+  assert.ok(labels.includes('Make'))
```
Comment (enmaku) [inline src/utils/exifFormat.test.js]: [blocker] These `labels.includes('Make')` / `'Date taken'` / `'Dimensions'` checks lock in user-facing copy. Per repo test rules, assert structure instead — e.g. row count, that `pickExifCore` fields appear in values, and that no row value contains GPS coordinates — without pinning panel labels that product copy may rename.

File: src/utils/exifFormat.test.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  })
+})
+
+test('pickExifCore returns null for null or non-object input', () => {
+  assert.equal(pickExifCore(null), null)
+  assert.equal(pickExifCore(undefined), null)
+  assert.equal(pickExifCore('not-an-object'), null)
+})
+
+test('formatExifTimestamp formats EXIF colon datetime', () => {
```
Comment (enmaku) [inline src/utils/exifFormat.test.js:36]: [structure] `formatExifTimestamp` uses `Intl.DateTimeFormat(undefined, …)`, so `/June|Jun/` varies by runtime locale. Prefer asserting non-empty output + year digits from the input, or pass an explicit locale to the formatter in test so the contract is stable.

File: src/composables/usePhotoGallery.js [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+  if (exifCache.has(src)) {
+    return exifCache.get(src)
+  }
+
+  try {
+    const data = await exifr.parse(src, {
+      tiff: true,
+      ifd0: true,
+      exif: true,
+      gps: true,
```
Comment (enmaku) [inline src/composables/usePhotoGallery.js]: [structure] `gps: true` still parses coordinates even though formatters and CONTEXT omit GPS from **photo details**. Set `gps: false` (or drop unused fields after parse) so privacy intent matches the parse cost.

File: src/composables/usePhotoGallery.js [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+    exifPanelOpen,
+    dialogExifRows,
+    isZoomed,
+    isDragging,
+    dialogImageWrap,
+    openPhoto,
+    toggleZoom,
+    onImagePointerDown,
+    onImageClick,
+    imageDragHandlers,
```
Comment (enmaku) [inline src/composables/usePhotoGallery.js]: [architecture] Exporting `imageDragHandlers` pushes `usePointerDragScroll` controller methods across the composable seam — callers must know drag-controller wiring. Wrap pointer move/end/cancel/leave inside the composable (same as `onImagePointerDown`) so `IndexPage` only binds composable-owned handlers.

File: src/composables/usePhotoGallery.js [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+      return
+    }
+
+    previewHistoryPushed.value = false
+    ignoreNextPopstate.value = true
+    history.back()
+    restoreGalleryScroll()
+  })
+
+  return {
```
Comment (enmaku) [inline src/composables/usePhotoGallery.js:313]: [architecture] The return surface is ~25 top-level bindings — interface nearly as wide as the implementation (shallow module). Group preview-related refs/handlers under one `preview` object (or similar) so the page destructures a few concepts instead of every ref.

File: src/pages/IndexPage.vue [hunk 6]
```diff
@@ -169,312 +169,33 @@
...
+  exifPanelOpen,
+  dialogExifRows,
+  isZoomed,
+  isDragging,
+  dialogImageWrap,
+  openPhoto,
+  toggleZoom,
+  onImagePointerDown,
+  onImageClick,
+  imageDragHandlers,
```
Comment (enmaku) [inline src/pages/IndexPage.vue]: [architecture] Destructuring `imageDragHandlers` from the composable keeps preview pan wiring in the page/template (`@pointermove="imageDragHandlers.onMove"`, …). Once wrapped in the composable, this export and the four template bindings can disappear.

**Review by enmaku — COMMENTED** [review]
## Verdict
COMMENT — would request changes on the tests, but GitHub disallows `REQUEST_CHANGES` on your own PR. The composable extraction is a clear win for `IndexPage.vue`; address the test-contract issues before merge.

## Thermo-nuclear (diff)
- **Blocker:** `exifFormat.test.js` asserts panel row labels (`Make`, `Date taken`, `Dimensions`, …) and formatted date month names. Those tests fail on copy edits with no logic change; rewrite around structure and privacy invariants (row count, field presence, GPS omission) instead of literal labels.
- **Structure:** `formatExifTimestamp` test depends on `Intl` locale output (`June|Jun`); CI/locale drift makes this flaky — assert parse success structurally or pin a locale in the formatter under test.
- **Minor:** `loadExif` still requests `gps: true` though formatted output never surfaces GPS — drop the parse flag to match the privacy contract and skip useless work.
- No file-size regression; `IndexPage` script slim-down is the right direction.

## Architecture (PR-scoped)
- **`usePhotoGallery` interface is shallow:** ~25 destructured bindings move complexity from the page into an equally wide composable seam — consider grouping preview state (dialog, zoom, exif panel, scroll restore) behind one object so callers hold fewer concepts.
- **`imageDragHandlers` leaks across the seam:** the page wires pointer lifecycle directly to the drag controller (template + destructuring); wrap move/end handlers inside the composable so callers don't depend on `usePointerDragScroll` internals.
- **Photo catalog + EXIF cache are module singletons:** fine for a single gallery route today; if a second consumer appears, extract a small `photoCatalog` module so the composable owns reactive UI state only.

## Unanchored notes
- PR title promises composable + EXIF tests; EXIF coverage is good, but history/back-navigation and zoom orchestration still have no tests through the composable interface. Not blocking this housekeeping PR, but worth a follow-up if preview regressions are a concern.
